tests/pytest_tests/unit_tests/test_lib/test_paths.py

```diff
@@ -0,0 +1,155 @@
+import os
+import platform
+from pathlib import PurePath, PurePosixPath, PureWindowsPath
+
+import pytest
+from hypothesis import example, given, settings
+from hypothesis.strategies import binary, one_of, sampled_from, text
+from hypothesis_fspaths import fspaths
+from wandb.sdk.lib.paths import (
+    PROHIBITED_CHARS,
+    RESERVED_NAMES,
+    LogicalPath,
+    sanitize_path,
+)
+from wandb.util import to_forward_slash_path
+
+
+@pytest.mark.parametrize(
+    "target,posix,windows,_bytes",
+    [
+        ("foo/bar.txt", "foo/bar.txt", "foo\\bar.txt", b"foo/bar.txt"),
+        ("et/tu.txt", "et/tu.txt", "et/tu.txt", b"et\\tu.txt"),
+        ("/ab/ra.txt", "/ab/ra.txt", "\\ab\\ra.txt", b"/ab/ra.txt"),
+        ("C:/a/b.txt", "C:/a/b.txt", "C:\\a\\b.txt", b"C:/a/b.txt"),
+    ],
+)
+def test_path_groups(target, posix, windows, _bytes):
+    assert target == LogicalPath(target)
+    assert target == LogicalPath(PurePosixPath(posix))
+    assert target == LogicalPath(PureWindowsPath(windows))
+    assert target == LogicalPath(_bytes) or platform.system() != "Windows"
```

Comment:
Do we not handle bytes properly on non-Windows platforms?

Reply:
Sort of. It's not really a bytes problem, it's that backslash is a permissible character on non-windows, so if you have `PurePosixPath("./\\/text.txt")` it will get stored that way. On the other hand, if you have `Path("./\\/text.txt")` then on Windows it will be sent to `test.txt` and on other platforms it will be sent to `\\/test.txt`. This is probably wrong, but changing it breaks current behavior. I'm torn on what we should actually do, but I figure that since this is a series of PRs for refactoring I should hold back on also introducing any behavior changes.